pkg/token/service.go

```diff
@@ -423,72 +438,105 @@ func (d *tokenService) updateToken(key CacheKey, tt mode) error {
+		return []error{fmt.Errorf("unable to create directory for tokens: %w", err)}
 	}
 
-	w := util.NewWriter()
+	var atErrorCount, rtErrorCount atomic.Int64
+	var wg sync.WaitGroup
+	echan := make(chan error, len(d.atTargetDomainRoles)+len(d.rtTargetDomainRoles))
 
-	err := d.accessTokenCache.Range(func(k CacheKey, t Token) error {
-		domain := t.Domain()
-		role := t.Role()
-		at := t.Raw()
-		if at == "" {
-			// skip placeholder token added during daemon creation
-			return nil
-		}
+	for _, k := range d.atTargetDomainRoles {
+		wg.Add(1)
+		go func(k CacheKey) {
+			defer wg.Done()
+			domain, role := k.Domain, k.Role
+			token := d.accessTokenCache.Load(k)
+			outPath := filepath.Join(d.tokenDir, domain+":role."+role+".accesstoken")
+			err := d.writeFileWithRetry(ctx, maxElapsedTime, token, outPath, mACCESS_TOKEN)
+			if err != nil {
+				echan <- err
+				atErrorCount.Add(1)
+			}
+		}(k)
+	}
 
-		log.Infof("[New Access Token] Domain: %s, Role: %s", domain, role)
-		outPath := filepath.Join(d.tokenDir, domain+":role."+role+".accesstoken")
-		log.Debugf("Saving Access Token[%d bytes] at %s", len(at), outPath)
-		if err := w.AddBytes(outPath, 0644, []byte(at)); err != nil {
-			return fmt.Errorf("unable to save access token: %w", err)
-		}
-		return nil
-	})
-	if err != nil {
-		return err
+	for _, k := range d.rtTargetDomainRoles {
+		wg.Add(1)
+		go func(k CacheKey) {
+			defer wg.Done()
+			domain, role := k.Domain, k.Role
+			token := d.roleTokenCache.Load(k)
+			outPath := filepath.Join(d.tokenDir, domain+":role."+role+".roletoken")
+			err := d.writeFileWithRetry(ctx, maxElapsedTime, token, outPath, mROLE_TOKEN)
+			if err != nil {
+				echan <- err
+				rtErrorCount.Add(1)
+			}
+		}(k)
 	}
-	err = d.roleTokenCache.Range(func(k CacheKey, t Token) error {
-		domain := t.Domain()
-		role := t.Role()
-		rt := t.Raw()
-		if rt == "" {
-			// skip placeholder token added during daemon creation
-			return nil
-		}
 
-		log.Infof("[New Role Token] Domain: %s, Role: %s", domain, role)
-		outPath := filepath.Join(d.tokenDir, domain+":role."+role+".roletoken")
-		log.Debugf("Saving Role Token[%d bytes] at %s", len(rt), outPath)
-		if err := w.AddBytes(outPath, 0644, []byte(rt)); err != nil {
-			return fmt.Errorf("unable to save role token: %w", err)
-		}
+	// wait for ALL token updates to complete
+	wg.Wait()
+	log.Infof("Saved tokens to files. accesstoken:success[%d],error[%d]; roletoken:success[%d],error[%d]", int64(len(d.atTargetDomainRoles))-atErrorCount.Load(), atErrorCount.Load(), int64(len(d.rtTargetDomainRoles))-rtErrorCount.Load(), rtErrorCount.Load())
+
+	// collect errors
+	close(echan)
+	errs := make([]error, 0, len(d.atTargetDomainRoles)+len(d.rtTargetDomainRoles))
+	for err := range echan {
+		errs = append(errs, err)
+	}
+	return errs
+}
+
+func (d *tokenService) writeFileWithRetry(ctx context.Context, maxElapsedTime time.Duration, token Token, outPath string, tt mode) error {
```

Comment:
Write a comment for what this does
i.e)
```
// writeFileWithRetry attempts multiple times to write a file for the given token (AT or RT) by utilizing writeFile()
```

Reply:
fixed in https://github.com/AthenZ/k8s-athenz-sia/pull/131/commits/6822928c8eef4cfb5ba60157b5141e6bcf66cfe1